feat: Add snowballs to lobby

Reapply opened the pull request.

Reapply committed:
Add LobbySnowballs feature to collect snowballs in lobby

Introduced a new LobbySnowballs class allowing players in the lobby region to collect snowballs from snow blocks while in adventure mode. This feature plays sound effects and spawns particles for an immersive experience. The LobbySnowballs is initialized in the main ChristmasEventPlugin.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/lobby/LobbySnowballs.kt:
```diff
@@ -0,0 +1,74 @@
+                playSnowCollectionEffects(player, clickedBlock.location)
+            }
+
+            // Cancel vanilla block breaking
+            isCancelled = true
```
The player would be in adventure; there's no need to cancel this, 
```suggestion
```

ShreyasAyyengar commented:
Do you have any images/videos/media to demonstrate the result of this PR?

ShreyasAyyengar commented:
In general, I'm OK with this PR, but I'm concerned about an entirely new class handling this, especially when this *could* be solved using the `object` feature of Koltin.

My recommendation is to just put the code inside of `HousekeepingEventListener` and nest the util methods inside the `event<...> {...}` code block to code clarity. 

Also, as a general rule, refrain from commenting obvious code. Unless something is unclear or incredibly miscellaneous something like:

```kotlin
// Only handle snow-type blocks
if (clickedBlock.type != Material.SNOW && clickedBlock.type != Material.SNOW_BLOCK) {
    return@event
}
```

or 

```kotlin
// Sound effects
player.playSound(location, Sound.BLOCK_SNOW_BREAK, 1.0f, 1.5f)
player.playSound(location, Sound.BLOCK_SNOW_STEP, 0.5f, 2.0f)
```

is unneeded, but I appreciate the clarity.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/lobby/LobbySnowballs.kt:
```diff
@@ -0,0 +1,74 @@
+        )
+    }
+
+    private fun giveSnowball(player: Player) {
+        player.inventory.addItem(ItemStack(Material.SNOWBALL, 1))
+    }
```
Avoid 1-line methods:

```suggestion
```

ShreyasAyyengar replied:
Avoid 1-line methods:

```suggestion
```

Reapply committed:
Consolidate snowball event handling logic.

Removed the LobbySnowballs class and relocated its functionality into HousekeepingEventListener for centralized event management. This change simplifies the codebase by eliminating duplicate logic and streamlining how the application handles player interactions with snow blocks.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/lobby/LobbySnowballs.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private fun shouldProcessEvent(event: PlayerInteractEvent): Boolean {
+        return eventController().currentGame == null &&
```
This is the only condition that will validate the event. The others are either implicitly known, or never changes (ex. gamemode)

By "this", I mean: 
```kotlin
        return eventController().currentGame == null &&
```

Since this change will reduce this method to one line, I would in-line all its uses.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/lobby/LobbySnowballs.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private fun giveSnowball(player: Player) {
+        player.inventory.addItem(ItemStack(Material.SNOWBALL, 1))
```
Avoid 1-line methods. 

```suggestion
private fun giveSnowball(player: Player) {
        player.inventory.addItem(ItemStack(Material.SNOWBALL, 1))
    }
```

Reapply committed:
Add quantity to snowball inventory addition

Specify a quantity of 1 when adding a snowball to the player's inventory to ensure clarity and prevent potential future issues. This update maintains consistency and enhances code readability.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/listeners/HousekeepingEventListener.kt:
```diff
@@ -312,4 +332,26 @@ class HousekeepingEventListener : Listener, PacketListener {
+                player.gameMode == GameMode.ADVENTURE
+    }
+
+    private fun playSnowCollectionEffects(player: Player, location: org.bukkit.Location) {
+        player.playSound(location, Sound.BLOCK_SNOW_BREAK, 1f, 1.5f)
+        player.playSound(location, Sound.BLOCK_SNOW_STEP, 0.5f, 2f)
+
+        location.world.spawnParticle(
+            org.bukkit.Particle.SNOWFLAKE,
+            location.add(0.5, 0.5, 0.5),
+            8, 0.2, 0.2, 0.2, 0.1
+        )
+    }
```
Nest this method under `event<PlayerInteractEvent> {...}`

Reapply committed:
Refactor playSnowCollectionEffects function into event

Moved the playSnowCollectionEffects function into the event block to streamline the code and ensure encapsulation. This change improves code readability and maintainability by limiting the function scope to where it is used.

Reapply committed:
Simplify snowball collection logic in HousekeepingEventListener

Removed unnecessary function definition for playing snow collection effects and streamlined conditional checks. The code now directly handles inventory updates and plays the associated sounds and particles more efficiently.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/listeners/HousekeepingEventListener.kt:
```diff
@@ -102,6 +107,29 @@ class HousekeepingEventListener : Listener, PacketListener {
+            val player = this.player
+            val clickedBlock = clickedBlock ?: return@event
+
+            if (eventController().currentGame != null ||
```
Invert this condition for readability: if current game is null OR clicked block is a snow variant, return.

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/listeners/HousekeepingEventListener.kt:
```diff
@@ -312,4 +340,16 @@ class HousekeepingEventListener : Listener, PacketListener {
+    )
+
+    private fun shouldProcessSnowballEvent(player: Player): Boolean {
+        return eventController().currentGame == null &&
```
This is the only condition that will validate the event. The others are either implicitly known, or never changes (ex. gamemode)

ShreyasAyyengar commented on src/main/kotlin/gg/flyte/christmas/listeners/HousekeepingEventListener.kt:
```diff
@@ -312,4 +340,16 @@ class HousekeepingEventListener : Listener, PacketListener {
+        MapSinglePoint(575, 120, 535)
+    )
+
+    private fun shouldProcessSnowballEvent(player: Player): Boolean {
```
Nest method under event codeblock.

Reapply committed:
Refactor snow interaction logic into separate functions

Separated the snow interaction logic into distinct functions to improve code organization and clarity. Functions `isSnowVariant` and `processSnowInteraction` now handle specific tasks, making the event handler cleaner and more focused. This change enhances readability and supports easier future modifications.

Reapply committed:
Inline snow interaction logic in event handler

Refactor the event listener by removing the processSnowInteraction function and directly inlining its logic within the PlayerInteractEvent handler. This consolidation simplifies the flow by handling snowball addition and effects directly within the event context, enhancing readability and maintainability.

ShreyasAyyengar committed:
fix: clean up code and use right reference.

Signed-off-by: Shreyas Ayyengar <shreyas.ayyengar@gmail.com>

ShreyasAyyengar commented:
[Thanks](https://github.com/flytegg/christmas-plugin-2024/commit/1cf9aafc06b95d760b40ea8a1ccb8be1921cbf0c).

Do send or PM me a Minecraft UUID so I can add you as an NPC in our contributors section as a thank you for your contribution.
